go/benchmarks/benchmarking.go

```diff
@@ -0,0 +1,418 @@
+}
+
+func keyFromNewKeyspace() string {
+	localRand := rand.New(rand.NewSource(time.Now().UnixNano()))
+	totalRange := sizeNewKeyspace - sizeExistingKeyspace
+	return fmt.Sprint(math.Floor(localRand.Float64()*float64(totalRange) + sizeExistingKeyspace + 1))
+}
+
+type actionLatency struct {
+	action  string
+	latency time.Duration
+}
+
+func runBenchmark(
+	iterationsPerTask int,
+	concurrentTasks int,
+	actions map[string]operations,
+	clients []benchmarkClient,
+) (totalDuration time.Duration, latencies map[string][]time.Duration) {
+	latencies = map[string][]time.Duration{
+		getExisting:    {},
+		getNonExisting: {},
+		set:            {},
+	}
+
+	numResults := concurrentTasks * iterationsPerTask
+	results := make(chan *actionLatency, numResults)
+	start := time.Now()
+	for i := 0; i < concurrentTasks; i++ {
+		go runTask(results, iterationsPerTask, actions, clients)
+	}
+
+	for i := 0; i < numResults; i++ {
+		result := <-results
+		latencies[result.action] = append(latencies[result.action], result.latency)
+	}
+
+	return time.Since(start), latencies
+}
+
+func runTask(results chan<- *actionLatency, iterations int, actions map[string]operations, clients []benchmarkClient) {
+	for i := 0; i < iterations; i++ {
+		clientIndex := i % len(clients)
+		action := randomAction()
+		operation := actions[action]
+		latency := measureOperation(operation, clients[clientIndex])
+		results <- &actionLatency{action: action, latency: latency}
+	}
+}
+
+func measureOperation(operation operations, client benchmarkClient) time.Duration {
+	start := time.Now()
+	_, err := operation(client)
+	duration := time.Since(start)
+	if err != nil {
+		log.Print("Error while executing operation: ", err)
+	}
+
+	return duration
+}
+
+const (
+	probGet            = 0.8
+	probGetExistingKey = 0.8
+)
+
+func randomAction() string {
+	localRand := rand.New(rand.NewSource(time.Now().UnixNano()))
+	if localRand.Float64() > probGet {
+		return set
+	}
+
+	if localRand.Float64() > probGetExistingKey {
+		return getNonExisting
+	}
+
+	return getExisting
+}
+
+type latencyStats struct {
+	avgLatency   time.Duration
+	p50Latency   time.Duration
+	p90Latency   time.Duration
+	p99Latency   time.Duration
+	stdDeviation time.Duration
+	numRequests  int
+}
+
+func getLatencyStats(actionLatencies map[string][]time.Duration) map[string]*latencyStats {
+	results := make(map[string]*latencyStats)
+
+	for action, latencies := range actionLatencies {
+		sort.Slice(latencies, func(i, j int) bool {
+			return latencies[i] < latencies[j]
+		})
+
+		results[action] = &latencyStats{
+			avgLatency:   average(latencies),
+			p50Latency:   percentile(latencies, 50),
+			p90Latency:   percentile(latencies, 90),
+			p99Latency:   percentile(latencies, 99),
+			stdDeviation: standardDeviation(latencies),
+			numRequests:  len(latencies),
+		}
+	}
+
+	return results
+}
+
+func average(observations []time.Duration) time.Duration {
```

Comment:
Is there a 3rd party go lib to calc stddev, avg and percentile?

Reply:
I'll try to find one